File: src/main/java/com/teamwizardry/librarianlib/ragdoll/cloth/PointMass3D.java
package com.teamwizardry.librarianlib.ragdoll.cloth;

import net.minecraft.util.math.Vec3d;

public class PointMass3D {
	public Vec3d pos, prevPos, origPos, friction;
	public float mass;
	public boolean pin = false;

	public PointMass3D(Vec3d pos, float mass) {
		this.prevPos = this.pos = pos;
		this.mass = mass;
	}
	
	public void applyForce(Vec3d force) {
		pos = pos.add(force.scale(1.0/mass));
	}
	
	public void applyMotion(Vec3d motion) {
		pos = pos.add(motion);
	}
}

File: src/main/java/com/teamwizardry/librarianlib/ragdoll/cloth/PointMass3D.kt
package com.teamwizardry.librarianlib.ragdoll.cloth

import net.minecraft.util.math.Vec3d

class PointMass3D(pos: Vec3d, var mass: Float) {
    var pos: Vec3d
    var prevPos: Vec3d
    var origPos: Vec3d? = null
    var friction: Vec3d? = null
    var pin = false

    init {
        this.prevPos = this.pos = pos
    }

    fun applyForce(force: Vec3d) {
        pos = pos.add(force.scale(1.0 / mass))
    }

    fun applyMotion(motion: Vec3d) {
        pos = pos.add(motion)
    }

    override fun toString(): String {
        return (if (pin) "[P]" else "") + pos.toString()
    }
}
